cpp/src/arrow/compute/kernels/scalar_temporal_test.cc

```diff
@@ -383,7 +384,96 @@ TEST_F(ScalarTemporalTest, DayOfWeek) {
+  const char* times_us_central = R"(["1969-12-31T18:00:00", null])";
+  auto options_utc = AssumeTimezoneOptions(timezone_utc);
+  auto options_kolkata = AssumeTimezoneOptions(timezone_kolkata);
+  auto options_us_central = AssumeTimezoneOptions(timezone_us_central);
+  auto options_invalid = AssumeTimezoneOptions("Europe/Brusselsss");
+
+  for (auto u : internal::AllTimeUnits()) {
+    auto unit = timestamp(u);
+    auto unit_utc = timestamp(u, timezone_utc);
+    auto unit_kolkata = timestamp(u, timezone_kolkata);
+    auto unit_us_central = timestamp(u, timezone_us_central);
+
+    CheckScalarUnary("assume_timezone", unit, times_utc, unit_utc, times_utc,
+                     &options_utc);
+    CheckScalarUnary("assume_timezone", unit, times_kolkata, unit_kolkata, times_utc,
+                     &options_kolkata);
+    CheckScalarUnary("assume_timezone", unit, times_us_central, unit_us_central,
+                     times_utc, &options_us_central);
+    ASSERT_RAISES(Invalid,
+                  AssumeTimezone(ArrayFromJSON(unit_kolkata, times_utc), options_utc));
+    ASSERT_RAISES(Invalid,
+                  AssumeTimezone(ArrayFromJSON(unit, times_utc), options_invalid));
+  }
+}
+
+TEST_F(ScalarTemporalTest, TestAssumeTimezoneAmbiguous) {
+  std::string timezone = "CET";
+  const char* times = R"(["2018-10-28 01:20:00",
+                          "2018-10-28 02:36:00",
+                          "2018-10-28 03:46:00"])";
+  const char* times_earliest = R"(["2018-10-27 23:20:00",
+                                   "2018-10-28 00:36:00",
+                                   "2018-10-28 02:46:00"])";
+  const char* times_latest = R"(["2018-10-27 23:20:00",
+                                 "2018-10-28 01:36:00",
+                                 "2018-10-28 02:46:00"])";
+
+  auto options_earliest = AssumeTimezoneOptions(
+      timezone, AssumeTimezoneOptions::Ambiguous::AMBIGUOUS_EARLIEST);
+  auto options_latest =
+      AssumeTimezoneOptions(timezone, AssumeTimezoneOptions::Ambiguous::AMBIGUOUS_LATEST);
+  auto options_raise =
+      AssumeTimezoneOptions(timezone, AssumeTimezoneOptions::Ambiguous::AMBIGUOUS_RAISE);
+
+  for (auto u : internal::AllTimeUnits()) {
+    auto unit = timestamp(u);
+    auto unit_local = timestamp(u, timezone);
+    ASSERT_RAISES(Invalid, AssumeTimezone(ArrayFromJSON(unit, times), options_raise));
+    CheckScalarUnary("assume_timezone", unit, times, unit_local, times_earliest,
+                     &options_earliest);
+    CheckScalarUnary("assume_timezone", unit, times, unit_local, times_latest,
+                     &options_latest);
+  }
+}
+
+TEST_F(ScalarTemporalTest, TestAssumeTimezoneNonexistent) {
+  std::string timezone = "Europe/Warsaw";
+  const char* times = R"(["2015-03-29 02:30:00", "2015-03-29 03:30:00"])";
+  const char* times_earliest = R"(["2015-03-29 01:00:00", "2015-03-29 01:30:00"])";
+  const char* times_latest = R"(["2015-03-29 01:00:00", "2015-03-29 01:30:00"])";
```

Comment:
Am I misreading, or is `times_latest` equal to `times_earliest`?

Reply:
I guess that the `date` library chooses the existing (UTC) instant that encloses the invalid instant "2015-03-29 02:30:00 Europe/Paris`. But having two option values that produce the same results doesn't seem useful at all.

Reply:
Yeah, indeed https://howardhinnant.github.io/date/tz.html#nonexistent_local_time

Reply:
I'll see if we can make it more like expected.

Reply:
Now the behavior is more like shift forward and shift backward.

Reply:
Uh, sorry, that's what not I meant :-) I think the behaviour was ok, it just that there shouldn't be two different option values that produce exactly the same result.

Reply:
Well maybe we can keep the original behavior as `infer` and have the two new ones as `shift_forward` and `shift_forward`. It needs a new name for sure.
I'm borrowing from [pandas tz_localize](https://pandas.pydata.org/docs/reference/api/pandas.Series.tz_localize.html).

Reply:
Well, that depends, do we want to copy Pandas here?
Look at the results in the Pandas example: "2015-03-29 01:59:59.999999999+01:00" and "2015-03-29 03:00:00+02:00" are separated by a single _nanosecond_. Is this behaviour useful? @jorisvandenbossche 

Also, note that `shift_forward` already corresponds to the original `date` library's behaviour. So we don't need an additional option values :-)

Reply:
> Look at the results in the Pandas example: "2015-03-29 01:59:59.999999999+01:00" and "2015-03-29 03:00:00+02:00" are separated by a single _nanosecond_. Is this behaviour useful? @jorisvandenbossche

I'm sure there are is someone out there who care about nanoseconds :)).

My thought here is - the input timestamp is half past full hour locally. The local hour doesn't exist, so the half past is probably a measurement mistake and what's was measured is half past an hour later or earlier. The third option (infer) is to ignore the half past and go to the moment the DST shift occurred.

```
earliest -> 2015-03-29 02:30:00 (Warsaw) -> 2015-03-29 00:30:00 (UTC)
infer -> 2015-03-29 02:30:00 (Warsaw) -> 2015-03-29 01:00:00 (UTC)
latest -> 2015-03-29 01:30:00 (Warsaw) -> 2015-03-29 01:30:00 (UTC)
```

Visual aid for the discussion (not exactly the same zones):
![](https://files.gitter.im/HowardHinnant/date/fM7T/EuropeMoscow.jpeg)

We can also just park this into [ARROW-13347](https://issues.apache.org/jira/browse/ARROW-13347).

Reply:
> My thought here is - the input timestamp is half past full hour locally. The local hour doesn't exist, so the half past is probably a measurement mistake and what was measured is half past an hour later or earlier. 

This corresponds neither to the Pandas behaviour nor to the `date` library behaviour. Is there any system that offers such an option? Saying "the half past is probably a measurement mistake and what was measured is half past an hour later or earlier" is quite a wild guess...

Reply:
Bottom line, I think we can simply implement the Pandas semantics while keeping the name "earliest" and "latest".

Reply:
> This corresponds neither to the Pandas behaviour nor to the `date` library behaviour. Is there any system that offers such an option? Saying "the half past is probably a measurement mistake and what was measured is half past an hour later or earlier" is quite a wild guess...

Well if the user wants to do it it's not guessing :)
Pandas is close with the `nonexistent=pd.Timedelta('1H')` but yeah, it is probably to niche to have here.

Reply:
> Bottom line, I think we can simply implement the Pandas semantics while keeping the name "earliest" and "latest".

Sounds good, will do.

Reply:
Pushed a pandas-like change. It's a bit annoying that output depends on the unit but ok ..

Reply:
Thank you!